Connect proxy failure if using driver of MySQL 8 and prepared statement

When I used driver `8.0.11` of MySQL and enabled prepared statement(using `&useServerPrepStmts=true&cachePrepStmts=true` in JDBC URL) to connect proxy, the exception occur:

```log
java.sql.SQLException: java.lang.ArrayIndexOutOfBoundsException: 1
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:127)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:95)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:87)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:61)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:71)
	at com.mysql.cj.jdbc.ServerPreparedStatement.executeInternal(ServerPreparedStatement.java:420)
	at com.mysql.cj.jdbc.ClientPreparedStatement.executeQuery(ClientPreparedStatement.java:1019)
	at Main.executeQuery(Main.java:51)
	at Main.main(Main.java:41)
Caused by: java.lang.ArrayIndexOutOfBoundsException: 1
	at com.mysql.cj.protocol.a.MergingColumnDefinitionFactory.createFromFields(MergingColumnDefinitionFactory.java:61)
	at com.mysql.cj.protocol.a.ColumnDefinitionReader.read(ColumnDefinitionReader.java:80)
	at com.mysql.cj.protocol.a.ColumnDefinitionReader.read(ColumnDefinitionReader.java:40)
	at com.mysql.cj.protocol.a.NativeProtocol.read(NativeProtocol.java:1648)
	at com.mysql.cj.protocol.a.BinaryResultsetReader.read(BinaryResultsetReader.java:70)
	at com.mysql.cj.protocol.a.BinaryResultsetReader.read(BinaryResultsetReader.java:50)
	at com.mysql.cj.protocol.a.NativeProtocol.read(NativeProtocol.java:1661)
	at com.mysql.cj.protocol.a.NativeProtocol.readAllResults(NativeProtocol.java:1707)
	at com.mysql.cj.ServerPreparedQuery.readExecuteResult(ServerPreparedQuery.java:419)
	at com.mysql.cj.ServerPreparedQuery.serverExecute(ServerPreparedQuery.java:202)
	at com.mysql.cj.jdbc.ServerPreparedStatement.serverExecute(ServerPreparedStatement.java:620)
	at com.mysql.cj.jdbc.ServerPreparedStatement.executeInternal(ServerPreparedStatement.java:397)
	... 3 more
```

But driver `5.1.42` of MySQL is fine.

The test class is:

```java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.text.DecimalFormat;

public final class Main {
    
    public static void main(final String[] args) throws ClassNotFoundException {
        String driverClassName = "com.mysql.jdbc.Driver";
        String url = "jdbc:mysql://localhost:3307/readwrite_splitting_db?serverTimezone=UTC&useSSL=false&useUnicode=true&characterEncoding=UTF-8&useServerPrepStmts=true&cachePrepStmts=true";
        String username = "root";
        String password = "root";
        String sql = "SELECT * FROM t_order WHERE order_id=?";
        Class.forName(driverClassName);
        try (Connection connection = DriverManager.getConnection(url, username, password);
             PreparedStatement preparedStatement = connection.prepareStatement(sql)
        ) {
            for (int i = 0; i < 10; i++) {
                executeQuery(preparedStatement);
            }
        } catch (final SQLException ex) {
            ex.printStackTrace();
        }
    }
    
    private static void executeQuery(final PreparedStatement preparedStatement) throws SQLException {
        preparedStatement.setLong(1, 1);
        try (ResultSet resultSet = preparedStatement.executeQuery()) {
            while (resultSet.next()) {
                resultSet.getObject(1);
            }
        }
    }
}

```

Configuration of config-readwrite-splitting.yaml

```yaml
schemaName: readwrite_splitting_db

dataSources:
  write_ds:
    url: jdbc:mysql://127.0.0.1:3306/demo_write_ds?serverTimezone=UTC&useSSL=false
    username: root
    password:
    connectionTimeoutMilliseconds: 30000
    idleTimeoutMilliseconds: 60000
    maxLifetimeMilliseconds: 1800000
    maxPoolSize: 50
    minPoolSize: 1
    maintenanceIntervalMilliseconds: 30000
  read_ds_0:
    url: jdbc:mysql://127.0.0.1:3306/demo_read_ds_0?serverTimezone=UTC&useSSL=false
    username: root
    password:
    connectionTimeoutMilliseconds: 30000
    idleTimeoutMilliseconds: 60000
    maxLifetimeMilliseconds: 1800000
    maxPoolSize: 50
    minPoolSize: 1
    maintenanceIntervalMilliseconds: 30000
  read_ds_1:
    url: jdbc:mysql://127.0.0.1:3306/demo_read_ds_1?serverTimezone=UTC&useSSL=false
    username: root
    password:
    connectionTimeoutMilliseconds: 30000
    idleTimeoutMilliseconds: 60000
    maxLifetimeMilliseconds: 1800000
    maxPoolSize: 50
    minPoolSize: 1
    maintenanceIntervalMilliseconds: 30000

rules:
- !READWRITE_SPLITTING
  dataSources:
    pr_ds:
      writeDataSourceName: write_ds
      readDataSourceNames:
        - read_ds_0
        - read_ds_1
```

Configuration of server.yaml

```yaml

rules:
  - !AUTHORITY
    users:
      - root@%:root
      - sharding@:sharding
    provider:
      type: NATIVE

props:
  max-connections-size-per-query: 1
  executor-size: 16  # Infinite by default.
  proxy-frontend-flush-threshold: 128  # The default value is 128.
    # LOCAL: Proxy will run with LOCAL transaction.
    # XA: Proxy will run with XA transaction.
    # BASE: Proxy will run with B.A.S.E transaction.
  proxy-transaction-type: LOCAL
  xa-transaction-manager-type: Atomikos
  proxy-opentracing-enabled: false
  proxy-hint-enabled: false
  sql-show: false
  check-table-metadata-enabled: false
  lock-wait-timeout-milliseconds: 50000 # The maximum time to wait for a lock
  show-process-list-enabled: false
```

--- a/shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/ProjectionsContext.java
+++ b/shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/ProjectionsContext.java
@@ -17,6 +17,7 @@
 
 package org.apache.shardingsphere.infra.binder.segment.select.projection;
 
+import com.google.common.base.Preconditions;
 import lombok.Getter;
 import lombok.RequiredArgsConstructor;
 import lombok.ToString;
@@ -24,6 +25,13 @@ import org.apache.shardingsphere.infra.binder.segment.select.projection.impl.Agg
 import org.apache.shardingsphere.infra.binder.segment.select.projection.impl.AggregationProjection;
 import org.apache.shardingsphere.infra.binder.segment.select.projection.impl.DerivedProjection;
 import org.apache.shardingsphere.infra.binder.segment.select.projection.impl.ShorthandProjection;
+import org.apache.shardingsphere.infra.metadata.ShardingSphereMetaData;
+import org.apache.shardingsphere.infra.metadata.schema.model.TableMetaData;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.item.ProjectionSegment;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.item.ProjectionsSegment;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.item.ShorthandProjectionSegment;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.generic.table.SimpleTableSegment;
+import org.apache.shardingsphere.sql.parser.sql.common.statement.dml.SelectStatement;
 
 import java.util.ArrayList;
 import java.util.Collection;
@@ -46,7 +54,7 @@ public final class ProjectionsContext {
     private final boolean distinctRow;
     
     private final Collection<Projection> projections;
-    
+
     /**
      * Judge is unqualified shorthand projection or not.
      * 
@@ -139,5 +147,32 @@ public final class ProjectionsContext {
             }
         }
         return result;
+    }
+
+    public static int getProjectionCount(ShardingSphereMetaData shardingSphereMetaData, SelectStatement selectStatement) {
+        ProjectionsSegment projections = selectStatement.getProjections();
+        Collection<ProjectionSegment> projectionSegments = projections.getProjections();
+        return isShorthandProjectionAndSimpleTable(projectionSegments, selectStatement)
+                ? getTableColumnCount(shardingSphereMetaData, (SimpleTableSegment) selectStatement.getFrom())
+                : projectionSegments.size();
+    }
+
+    private static boolean isShorthandProjectionAndSimpleTable(Collection<ProjectionSegment> c, SelectStatement statement) {
+        return isShorthandProjection(c) && isSimpleTableStatement(statement);
+    }
+
+    private static boolean isShorthandProjection(Collection<ProjectionSegment> c) {
+        return c.size() == 1 && c.iterator().next() instanceof ShorthandProjectionSegment;
+    }
+
+    private static boolean isSimpleTableStatement(SelectStatement statement) {
+        return statement.getFrom() instanceof SimpleTableSegment;
+    }
+
+    private static int getTableColumnCount(ShardingSphereMetaData shardingSphereMetaData, SimpleTableSegment table) {
+        String tableName = table.getTableName().getIdentifier().getValue();
+        TableMetaData tableMetaData = shardingSphereMetaData.getSchema().get(tableName);
+        Preconditions.checkNotNull(tableMetaData, "Table '%s' not found", tableName);
+        return tableMetaData.getColumns().size();
     }
 }

--- a/shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/ProjectionsContext.java
+++ b/shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/ProjectionsContext.java
@@ -17,7 +17,6 @@
 
 package org.apache.shardingsphere.infra.binder.segment.select.projection;
 
-import com.google.common.base.Preconditions;
 import lombok.Getter;
 import lombok.RequiredArgsConstructor;
 import lombok.ToString;
@@ -25,13 +24,6 @@ import org.apache.shardingsphere.infra.binder.segment.select.projection.impl.Agg
 import org.apache.shardingsphere.infra.binder.segment.select.projection.impl.AggregationProjection;
 import org.apache.shardingsphere.infra.binder.segment.select.projection.impl.DerivedProjection;
 import org.apache.shardingsphere.infra.binder.segment.select.projection.impl.ShorthandProjection;
-import org.apache.shardingsphere.infra.metadata.ShardingSphereMetaData;
-import org.apache.shardingsphere.infra.metadata.schema.model.TableMetaData;
-import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.item.ProjectionSegment;
-import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.item.ProjectionsSegment;
-import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.item.ShorthandProjectionSegment;
-import org.apache.shardingsphere.sql.parser.sql.common.segment.generic.table.SimpleTableSegment;
-import org.apache.shardingsphere.sql.parser.sql.common.statement.dml.SelectStatement;
 
 import java.util.ArrayList;
 import java.util.Collection;
@@ -149,30 +141,16 @@ public final class ProjectionsContext {
         return result;
     }
 
-    public static int getProjectionCount(ShardingSphereMetaData shardingSphereMetaData, SelectStatement selectStatement) {
-        ProjectionsSegment projections = selectStatement.getProjections();
-        Collection<ProjectionSegment> projectionSegments = projections.getProjections();
-        return isShorthandProjectionAndSimpleTable(projectionSegments, selectStatement)
-                ? getTableColumnCount(shardingSphereMetaData, (SimpleTableSegment) selectStatement.getFrom())
-                : projectionSegments.size();
+    public int getProjectionCount() {
+        return projections.stream()
+                .mapToInt(this::getActualColumnCount)
+                .sum();
     }
 
-    private static boolean isShorthandProjectionAndSimpleTable(Collection<ProjectionSegment> c, SelectStatement statement) {
-        return isShorthandProjection(c) && isSimpleTableStatement(statement);
+    private int getActualColumnCount(Projection projection) {
+        return projection instanceof ShorthandProjection
+                ? ((ShorthandProjection) projection).getActualColumns().size()
+                : 1;
     }
 
-    private static boolean isShorthandProjection(Collection<ProjectionSegment> c) {
-        return c.size() == 1 && c.iterator().next() instanceof ShorthandProjectionSegment;
-    }
-
-    private static boolean isSimpleTableStatement(SelectStatement statement) {
-        return statement.getFrom() instanceof SimpleTableSegment;
-    }
-
-    private static int getTableColumnCount(ShardingSphereMetaData shardingSphereMetaData, SimpleTableSegment table) {
-        String tableName = table.getTableName().getIdentifier().getValue();
-        TableMetaData tableMetaData = shardingSphereMetaData.getSchema().get(tableName);
-        Preconditions.checkNotNull(tableMetaData, "Table '%s' not found", tableName);
-        return tableMetaData.getColumns().size();
-    }
 }

--- a/shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/ProjectionsContext.java
+++ b/shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/ProjectionsContext.java
@@ -38,18 +38,18 @@ import java.util.Optional;
 @Getter
 @ToString
 public final class ProjectionsContext {
-    
+
     private final int startIndex;
-    
+
     private final int stopIndex;
-    
+
     private final boolean distinctRow;
-    
+
     private final Collection<Projection> projections;
 
     /**
      * Judge is unqualified shorthand projection or not.
-     * 
+     *
      * @return is unqualified shorthand projection or not
      */
     public boolean isUnqualifiedShorthandProjection() {
@@ -59,10 +59,10 @@ public final class ProjectionsContext {
         Projection projection = projections.iterator().next();
         return projection instanceof ShorthandProjection && !((ShorthandProjection) projection).getOwner().isPresent();
     }
-    
+
     /**
      * Find alias.
-     * 
+     *
      * @param projectionName projection name
      * @return projection alias
      */
@@ -74,7 +74,7 @@ public final class ProjectionsContext {
         }
         return Optional.empty();
     }
-    
+
     /**
      * Find projection index.
      *
@@ -91,7 +91,7 @@ public final class ProjectionsContext {
         }
         return Optional.empty();
     }
-    
+
     /**
      * Get aggregation projections.
      *
@@ -108,7 +108,7 @@ public final class ProjectionsContext {
         }
         return result;
     }
-    
+
     /**
      * Get aggregation distinct projections.
      *
@@ -123,10 +123,10 @@ public final class ProjectionsContext {
         }
         return result;
     }
-    
+
     /**
      * Get expand projections with shorthand projections.
-     * 
+     *
      * @return expand projections
      */
     public List<Projection> getExpandProjections() {
@@ -139,18 +139,6 @@ public final class ProjectionsContext {
             }
         }
         return result;
-    }
-
-    public int getProjectionCount() {
-        return projections.stream()
-                .mapToInt(this::getActualColumnCount)
-                .sum();
-    }
-
-    private int getActualColumnCount(Projection projection) {
-        return projection instanceof ShorthandProjection
-                ? ((ShorthandProjection) projection).getActualColumns().size()
-                : 1;
     }
 
 }

--- a/shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java
+++ b/shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java
@@ -82,7 +82,7 @@ public final class MySQLComStmtPrepareExecutor implements CommandExecutor {
             SelectStatementContext sqlStatementContext = (SelectStatementContext) SQLStatementContextFactory.newInstance(
                     metaDataMap, Collections.emptyList(), sqlStatement, schemaName);
             ProjectionsContext projectionsContext = sqlStatementContext.getProjectionsContext();
-            projectionCount = projectionsContext.getProjectionCount();
+            projectionCount = projectionsContext.getExpandProjections().size();
         }
         return projectionCount;
     }

--- a/shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java
+++ b/shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java
@@ -26,6 +26,7 @@ import org.apache.shardingsphere.db.protocol.mysql.packet.command.query.binary.p
 import org.apache.shardingsphere.db.protocol.mysql.packet.command.query.binary.prepare.MySQLComStmtPreparePacket;
 import org.apache.shardingsphere.db.protocol.mysql.packet.generic.MySQLEofPacket;
 import org.apache.shardingsphere.db.protocol.packet.DatabasePacket;
+import org.apache.shardingsphere.infra.binder.segment.select.projection.ProjectionsContext;
 import org.apache.shardingsphere.infra.database.type.DatabaseTypeRegistry;
 import org.apache.shardingsphere.infra.metadata.ShardingSphereMetaData;
 import org.apache.shardingsphere.infra.metadata.schema.model.TableMetaData;
@@ -73,37 +74,13 @@ public final class MySQLComStmtPrepareExecutor implements CommandExecutor {
     }
     
     private int getProjectionCount(final SQLStatement sqlStatement) {
-        return sqlStatement instanceof SelectStatement
-                ? getProjectionCount((SelectStatement) sqlStatement) : 0;
-    }
-
-    private int getProjectionCount(SelectStatement selectStatement) {
-        ProjectionsSegment projections = selectStatement.getProjections();
-        Collection<ProjectionSegment> projectionSegments = projections.getProjections();
-        return isShorthandProjectionAndSimpleTable(projectionSegments, selectStatement)
-            ? getTableColumnCount((SimpleTableSegment) selectStatement.getFrom())
-            : projectionSegments.size();
-    }
-
-    private boolean isShorthandProjectionAndSimpleTable(Collection<ProjectionSegment> c, SelectStatement statement) {
-        return isShorthandProjection(c) && isSimpleTableStatement(statement);
-    }
-
-    private boolean isShorthandProjection(Collection<ProjectionSegment> c) {
-        return c.size() == 1 && c.iterator().next() instanceof ShorthandProjectionSegment;
-    }
-
-    private boolean isSimpleTableStatement(SelectStatement statement) {
-        return statement.getFrom() instanceof SimpleTableSegment;
-    }
-
-    private int getTableColumnCount(SimpleTableSegment table) {
-        String schemaName = backendConnection.getSchemaName();
-        ShardingSphereMetaData shardingSphereMetaData = ProxyContext.getInstance().getMetaData(schemaName);
-        String tableName = table.getTableName().getIdentifier().getValue();
-        TableMetaData tableMetaData = shardingSphereMetaData.getSchema().get(tableName);
-        Preconditions.checkNotNull(tableMetaData, "Table '%s' not found", tableName);
-        return tableMetaData.getColumns().size();
+        int projectionCount = 0;
+        if (sqlStatement instanceof SelectStatement) {
+            String schemaName = backendConnection.getSchemaName();
+            ShardingSphereMetaData shardingSphereMetaData = ProxyContext.getInstance().getMetaData(schemaName);
+            projectionCount = ProjectionsContext.getProjectionCount(shardingSphereMetaData, (SelectStatement) sqlStatement);
+        }
+        return projectionCount;
     }
 
     private Collection<DatabasePacket<?>> createPackets(final int statementId, final int projectionCount, final int parameterCount) {

--- a/shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java
+++ b/shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java
@@ -100,11 +100,9 @@ public final class MySQLComStmtPrepareExecutor implements CommandExecutor {
     private int getTableColumnCount(SimpleTableSegment table) {
         String schemaName = backendConnection.getSchemaName();
         ShardingSphereMetaData shardingSphereMetaData = ProxyContext.getInstance().getMetaData(schemaName);
-
         String tableName = table.getTableName().getIdentifier().getValue();
         TableMetaData tableMetaData = shardingSphereMetaData.getSchema().get(tableName);
         Preconditions.checkNotNull(tableMetaData, "Table '%s' not found", tableName);
-
         return tableMetaData.getColumns().size();
     }
 

--- a/shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java
+++ b/shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java
@@ -17,7 +17,6 @@
 
 package org.apache.shardingsphere.proxy.frontend.mysql.command.query.binary.prepare;
 
-import com.google.common.base.Preconditions;
 import lombok.RequiredArgsConstructor;
 import org.apache.shardingsphere.db.protocol.mysql.constant.MySQLBinaryColumnType;
 import org.apache.shardingsphere.db.protocol.mysql.packet.command.query.MySQLColumnDefinition41Packet;
@@ -29,16 +28,11 @@ import org.apache.shardingsphere.db.protocol.packet.DatabasePacket;
 import org.apache.shardingsphere.infra.binder.segment.select.projection.ProjectionsContext;
 import org.apache.shardingsphere.infra.database.type.DatabaseTypeRegistry;
 import org.apache.shardingsphere.infra.metadata.ShardingSphereMetaData;
-import org.apache.shardingsphere.infra.metadata.schema.model.TableMetaData;
 import org.apache.shardingsphere.infra.parser.ShardingSphereSQLParserEngine;
 import org.apache.shardingsphere.proxy.backend.communication.jdbc.connection.BackendConnection;
 import org.apache.shardingsphere.proxy.backend.context.ProxyContext;
 import org.apache.shardingsphere.proxy.frontend.command.executor.CommandExecutor;
 import org.apache.shardingsphere.proxy.frontend.exception.UnsupportedPreparedStatementException;
-import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.item.ProjectionSegment;
-import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.item.ProjectionsSegment;
-import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.item.ShorthandProjectionSegment;
-import org.apache.shardingsphere.sql.parser.sql.common.segment.generic.table.SimpleTableSegment;
 import org.apache.shardingsphere.sql.parser.sql.common.statement.SQLStatement;
 import org.apache.shardingsphere.sql.parser.sql.common.statement.dml.SelectStatement;
 

--- a/shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java
+++ b/shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java
@@ -17,6 +17,7 @@
 
 package org.apache.shardingsphere.proxy.frontend.mysql.command.query.binary.prepare;
 
+import com.google.common.base.Preconditions;
 import lombok.RequiredArgsConstructor;
 import org.apache.shardingsphere.db.protocol.mysql.constant.MySQLBinaryColumnType;
 import org.apache.shardingsphere.db.protocol.mysql.packet.command.query.MySQLColumnDefinition41Packet;
@@ -26,11 +27,17 @@ import org.apache.shardingsphere.db.protocol.mysql.packet.command.query.binary.p
 import org.apache.shardingsphere.db.protocol.mysql.packet.generic.MySQLEofPacket;
 import org.apache.shardingsphere.db.protocol.packet.DatabasePacket;
 import org.apache.shardingsphere.infra.database.type.DatabaseTypeRegistry;
+import org.apache.shardingsphere.infra.metadata.ShardingSphereMetaData;
+import org.apache.shardingsphere.infra.metadata.schema.model.TableMetaData;
 import org.apache.shardingsphere.infra.parser.ShardingSphereSQLParserEngine;
 import org.apache.shardingsphere.proxy.backend.communication.jdbc.connection.BackendConnection;
 import org.apache.shardingsphere.proxy.backend.context.ProxyContext;
 import org.apache.shardingsphere.proxy.frontend.command.executor.CommandExecutor;
 import org.apache.shardingsphere.proxy.frontend.exception.UnsupportedPreparedStatementException;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.item.ProjectionSegment;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.item.ProjectionsSegment;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.item.ShorthandProjectionSegment;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.generic.table.SimpleTableSegment;
 import org.apache.shardingsphere.sql.parser.sql.common.statement.SQLStatement;
 import org.apache.shardingsphere.sql.parser.sql.common.statement.dml.SelectStatement;
 
@@ -66,9 +73,41 @@ public final class MySQLComStmtPrepareExecutor implements CommandExecutor {
     }
     
     private int getProjectionCount(final SQLStatement sqlStatement) {
-        return sqlStatement instanceof SelectStatement ? ((SelectStatement) sqlStatement).getProjections().getProjections().size() : 0;
+        return sqlStatement instanceof SelectStatement
+                ? getProjectionCount((SelectStatement) sqlStatement) : 0;
     }
-    
+
+    private int getProjectionCount(SelectStatement selectStatement) {
+        ProjectionsSegment projections = selectStatement.getProjections();
+        Collection<ProjectionSegment> projectionSegments = projections.getProjections();
+        return isShorthandProjectionAndSimpleTable(projectionSegments, selectStatement)
+            ? getTableColumnCount((SimpleTableSegment) selectStatement.getFrom())
+            : projectionSegments.size();
+    }
+
+    private boolean isShorthandProjectionAndSimpleTable(Collection<ProjectionSegment> c, SelectStatement statement) {
+        return isShorthandProjection(c) && isSimpleTableStatement(statement);
+    }
+
+    private boolean isShorthandProjection(Collection<ProjectionSegment> c) {
+        return c.size() == 1 && c.iterator().next() instanceof ShorthandProjectionSegment;
+    }
+
+    private boolean isSimpleTableStatement(SelectStatement statement) {
+        return statement.getFrom() instanceof SimpleTableSegment;
+    }
+
+    private int getTableColumnCount(SimpleTableSegment table) {
+        String schemaName = backendConnection.getSchemaName();
+        ShardingSphereMetaData shardingSphereMetaData = ProxyContext.getInstance().getMetaData(schemaName);
+
+        String tableName = table.getTableName().getIdentifier().getValue();
+        TableMetaData tableMetaData = shardingSphereMetaData.getSchema().get(tableName);
+        Preconditions.checkNotNull(tableMetaData, "Table '%s' not found", tableName);
+
+        return tableMetaData.getColumns().size();
+    }
+
     private Collection<DatabasePacket<?>> createPackets(final int statementId, final int projectionCount, final int parameterCount) {
         Collection<DatabasePacket<?>> result = new LinkedList<>();
         result.add(new MySQLComStmtPrepareOKPacket(++currentSequenceId, statementId, projectionCount, parameterCount, 0));

--- a/shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java
+++ b/shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/prepare/MySQLComStmtPrepareExecutor.java
@@ -25,7 +25,10 @@ import org.apache.shardingsphere.db.protocol.mysql.packet.command.query.binary.p
 import org.apache.shardingsphere.db.protocol.mysql.packet.command.query.binary.prepare.MySQLComStmtPreparePacket;
 import org.apache.shardingsphere.db.protocol.mysql.packet.generic.MySQLEofPacket;
 import org.apache.shardingsphere.db.protocol.packet.DatabasePacket;
+import org.apache.shardingsphere.infra.binder.SQLStatementContextFactory;
 import org.apache.shardingsphere.infra.binder.segment.select.projection.ProjectionsContext;
+import org.apache.shardingsphere.infra.binder.statement.dml.SelectStatementContext;
+import org.apache.shardingsphere.infra.context.metadata.MetaDataContexts;
 import org.apache.shardingsphere.infra.database.type.DatabaseTypeRegistry;
 import org.apache.shardingsphere.infra.metadata.ShardingSphereMetaData;
 import org.apache.shardingsphere.infra.parser.ShardingSphereSQLParserEngine;
@@ -37,7 +40,9 @@ import org.apache.shardingsphere.sql.parser.sql.common.statement.SQLStatement;
 import org.apache.shardingsphere.sql.parser.sql.common.statement.dml.SelectStatement;
 
 import java.util.Collection;
+import java.util.Collections;
 import java.util.LinkedList;
+import java.util.Map;
 
 /**
  * COM_STMT_PREPARE command executor for MySQL.
@@ -71,8 +76,13 @@ public final class MySQLComStmtPrepareExecutor implements CommandExecutor {
         int projectionCount = 0;
         if (sqlStatement instanceof SelectStatement) {
             String schemaName = backendConnection.getSchemaName();
-            ShardingSphereMetaData shardingSphereMetaData = ProxyContext.getInstance().getMetaData(schemaName);
-            projectionCount = ProjectionsContext.getProjectionCount(shardingSphereMetaData, (SelectStatement) sqlStatement);
+            ProxyContext proxyContext = ProxyContext.getInstance();
+            MetaDataContexts metaDataContexts = proxyContext.getMetaDataContexts();
+            Map<String, ShardingSphereMetaData> metaDataMap = metaDataContexts.getMetaDataMap();
+            SelectStatementContext sqlStatementContext = (SelectStatementContext) SQLStatementContextFactory.newInstance(
+                    metaDataMap, Collections.emptyList(), sqlStatement, schemaName);
+            ProjectionsContext projectionsContext = sqlStatementContext.getProjectionsContext();
+            projectionCount = projectionsContext.getProjectionCount();
         }
         return projectionCount;
     }

